sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -208,6 +210,78 @@ case class Abs(child: Expression, failOnError: Boolean = SQLConf.get.ansiEnabled
   override protected def withNewChildInternal(newChild: Expression): Abs = copy(child = newChild)
 }
 
+/**
+ * The child class should override decimalType method to report the result data type.
+ *
+ * When `spark.sql.decimalOperations.allowPrecisionLoss` is set to true, if the precision / scale
+ * needed are out of the range of available values, the scale is reduced up to 6, in order to
+ * prevent the truncation of the integer part of the decimals.
+ *
+ * Rounds the decimal to given scale and check whether the decimal can fit in provided precision
+ * or not. If not, if `nullOnOverflow` is `true`, it returns `null`; otherwise an
+ * `ArithmeticException` is thrown.
+ */
+trait DecimalArithmeticSupport extends BinaryArithmetic {
+  protected val nullOnOverflow: Boolean = !failOnError
+  protected val allowPrecisionLoss: Boolean = SQLConf.get.decimalOperationsAllowPrecisionLoss
+
+  override def checkInputDataTypes(): TypeCheckResult = (left.dataType, right.dataType) match {
+    case (_: DecimalType, _: DecimalType) =>
+      // We allow eval decimal type with different precision and scale, and change the precision
+      // and scale before return result.
+      TypeCheckResult.TypeCheckSuccess
+    case _ => super.checkInputDataTypes()
+  }
+
+  /** Name of the function for this expression on a [[Decimal]] type. */
+  protected def decimalMethod: String =
+    throw QueryExecutionErrors.notOverrideExpectedMethodsError("DecimalArithmetic",
+      "decimalMethod", "genCode")
+  protected def decimalType(p1: Int, s1: Int, p2: Int, s2: Int): DecimalType =
+    throw QueryExecutionErrors.notOverrideExpectedMethodsError("DecimalArithmetic",
+      "decimalType", "dataType")
+
+  override def nullable: Boolean = dataType match {
+    case _: DecimalType => true
```

Comment:
should it depend on `nullOnOverflow`?

Reply:
yes, nullOnOverflow is more accurate

Reply:
after some thought, we can not simply use `nullOnOverflow` here. If we want to use `nullOnOverflow`, the code-gen should be:
```
if (nullOnOverflow) {
  ${ev.isNull} = ${ev.value} == null;
}
```